In [1]:
import sys
import os

project_root = os.path.abspath('..')
if project_root not in sys.path:
    sys.path.append(project_root)

%load_ext autoreload
%autoreload 2


In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import warnings

from src.features import build_features

warnings.filterwarnings("ignore")
sns.set_style()
plt.rcParams["figure.figsize"] = (12, 6)

HOSPITAL_DEL_TORAX = 112103

## 1. Análisis de Procedimientos a Nivel País

En este análisis se quieren responder las siguiente preguntas:

1. ¿Cómo es el ranking de los hospitales para cada uno de los procedimientos realizads?
2. ¿Quién es el Hospital que realiza más de X procedimientos?
3. ¿El procedimiento X es realizado en todo el país? ¿En qué partes en específico se realiza?

Para responder esta pregunta se utilizará la base de datos de Egresos de Hospitales adscritos
a GRD entre 2019 y 2021.


In [3]:
vars_procedimientos = {f"PROCEDIMIENTO{i}": str for i in range(1, 31)}

In [4]:
df_procesada = pd.read_csv("../data/processed/df_procesada.csv", sep=";", dtype=vars_procedimientos)
df_procesada["ANIO_EGRESO"] = df_procesada["ANIO_EGRESO"].astype("Int32")

In [5]:
display(df_procesada.head(10))

COD_HOSPITAL  CIP_ENCRIPTADO    SEXO FECHA_NACIMIENTO  ETNIA   PROVINCIA  \
0        118100       1314867.0   MUJER       1988-06-19  OTRO   CONCEPCION   
1        118100       1354418.0   MUJER       1988-03-27  OTRO   CONCEPCION   
2        118100        239861.0   MUJER       2017-06-17  OTRO   CONCEPCION   
3        118100        330326.0  HOMBRE       2017-12-04  OTRO   CONCEPCION   
4        118100       1369293.0   MUJER       1995-12-06  OTRO   CONCEPCION   
5        118100        469385.0   MUJER       1942-06-29  OTRO   CONCEPCION   
6        118100       1146635.0   MUJER       1951-07-14  OTRO   CONCEPCION   
7        118100         93557.0  HOMBRE       1938-03-04  OTRO   CONCEPCION   
8        118100        674943.0   MUJER       1962-08-21  OTRO   CONCEPCION   
9        118100       1079955.0  HOMBRE       1944-12-07  OTRO   CONCEPCION   

                COMUNA NACIONALIDAD                       PREVISION  \
0  SAN PEDRO DE LA PAZ        CHILE  FONASA INSTITUCIONAL - (MAI) A   
1  SAN PEDRO DE LA PAZ        CHILE  FONASA INSTITUCIONAL - (MAI) D   
2           CONCEPCIÓN        CHILE  FONASA INSTITUCIONAL - (MAI) B   
3           CONCEPCIÓN        CHILE  FONASA INSTITUCIONAL - (MAI) A   
4              FLORIDA        CHILE  FONASA INSTITUCIONAL - (MAI) C   
5          CHIGUAYANTE        CHILE  FONASA INSTITUCIONAL - (MAI) B   
6           CONCEPCIÓN        CHILE  FONASA INSTITUCIONAL - (MAI) B   
7                 LOTA        CHILE  FONASA INSTITUCIONAL - (MAI) D   
8          SANTA JUANA        CHILE  FONASA INSTITUCIONAL - (MAI) B   
9           CONCEPCIÓN        CHILE  FONASA INSTITUCIONAL - (MAI) B   

  SERVICIO_SALUD  ... USOSPABELLON IR_29301_COD_GRD IR_29301_PESO  \
0     CONCEPCIÓN  ...          1.0         146102.0        0.5744   
1     CONCEPCIÓN  ...          NaN         134161.0        0.2951   
2     CONCEPCIÓN  ...          NaN          64182.0        0.6736   
3     CONCEPCIÓN  ...          NaN          64172.0        0.5475   
4     CONCEPCIÓN  ...          NaN         134141.0        0.3107   
5     CONCEPCIÓN  ...          NaN         114103.0        1.1799   
6     CONCEPCIÓN  ...          NaN          44163.0        1.0328   
7     CONCEPCIÓN  ...          1.0          11101.0        1.8731   
8     CONCEPCIÓN  ...          1.0          61402.0        1.3078   
9     CONCEPCIÓN  ...          NaN          14133.0        1.6687   

  IR_29301_SEVERIDAD IR_29301_MORTALIDAD  \
0                2.0                 1.0   
1                1.0                 1.0   
2                2.0                 2.0   
3                2.0                 2.0   
4                1.0                 1.0   
5                3.0                 3.0   
6                3.0                 3.0   
7                1.0                 1.0   
8                2.0                 2.0   
9                3.0                 3.0   

                          HOSPPROCEDENCIA ESTANCIA ANIO_EGRESO MES_EGRESO  \
0                                     NaN      5.0        2019        3.0   
1                                     NaN      4.0        2019        3.0   
2                                     NaN      6.0        2019        3.0   
3                                     NaN      2.0        2019        3.0   
4                                     NaN      6.0        2019        3.0   
5                                     NaN     47.0        2019        3.0   
6                                     NaN     43.0        2019        3.0   
7                                     NaN     41.0        2019        3.0   
8  HOSPITAL CLORINDA AVELLO (SANTA JUANA)     33.0        2019        3.0   
9                                     NaN     31.0        2019        3.0   

    FECHA  
0  2019-3  
1  2019-3  
2  2019-3  
3  2019-3  
4  2019-3  
5  2019-3  
6  2019-3  
7  2019-3  
8  2019-3  
9  2019-3  

[10 rows x 133 columns]

Los procedimientos en esta base de datos se encuentran en las columnas "PROCEDIMIENTOSX", donde X
va desde 1 hasta 30. Para analizar esta base de forma más expedita, se transformará de formato
ancho a formato largo, utilizando como identificador a las columnas ["COD_HOSPITAL", "ANIO_EGRESO",
"CIP_ENCRIPTADO", "DIAGNOSTICO1"]. Cabe destacar que la columna "CIP_ENCRIPTADO" hace referencia al
RUT del paciente.


In [6]:
procedimientos_formato_long = pd.melt(
    df_procesada,
    id_vars=["COD_HOSPITAL", "ANIO_EGRESO", "CIP_ENCRIPTADO", "DIAGNOSTICO1"],
    value_vars=vars_procedimientos.keys(),
    value_name="procedimiento",
)

In [7]:
conteo_y_ranking_procedimientos = build_features.contar_procedimientos_y_rankear(
    procedimientos_formato_long
)

In [8]:
display(conteo_y_ranking_procedimientos.head(20))

ANIO_EGRESO  COD_HOSPITAL procedimiento  conteo  ranking  \
180193         2021        113180         99.99       4        1   
160738         2021        108100         99.99       2        2   
148451         2021        103100         99.99       1        3   
152391         2021        105100         99.99       1        4   
166209         2021        110100         99.99       1        5   
167455         2021        110120         99.99       1        6   
179376         2021        113130         99.99       1        7   
182902         2021        114101         99.99       1        8   
187970         2021        115107         99.99       1        9   
188758         2021        115110         99.99       1       10   
197375         2021        118100         99.99       1       11   
202548         2021        121109         99.99       1       12   
206349         2021        122100         99.99       1       13   
211747         2021        126100         99.99       1       14   
213212         2021        129106         99.98      47        1   
206598         2021        123100         99.98      25        2   
162279         2021        109100         99.98      16        3   
180013         2021        113180         99.98      15        4   
165111         2021        110100         99.98      14        5   
196061         2021        118100         99.98      14        6   

        diferencia_proximo_hospital  total_procedimientos  
180193                            2                    18  
160738                            1                    18  
148451                            0                    18  
152391                            0                    18  
166209                            0                    18  
167455                            0                    18  
179376                            0                    18  
182902                            0                    18  
187970                            0                    18  
188758                            0                    18  
197375                            0                    18  
202548                            0                    18  
206349                            0                    18  
211747                          -46                    18  
213212                           22                   235  
206598                            9                   235  
162279                            1                   235  
180013                            1                   235  
165111                            0                   235  
196061                            0                   235

El resultado muestra el ranking para todos los procedimientos asignados en el país en cada uno de 
los años analizados (2019 a 2021). 

A modo de ejemplo, se ilustra el ranking en el año 2021 para el procedimiento 99.99. En este caso,
se observa que los hospitales con código 108100 y 113180 están en los primeros lugares, realizando
2 procedimientos cada uno.

Ahora, si se filtra por el hospital del tórax, y sólo los procedimientos donde el hospital sea
ranking 1, entonces se obtendrán procedimientos altamente relevantes para el país.

In [9]:
mejores_procedimientos_torax = conteo_y_ranking_procedimientos.query(
    "COD_HOSPITAL == @HOSPITAL_DEL_TORAX and ranking == 1"
).sort_values(["ANIO_EGRESO", "conteo"],ascending=False)

In [10]:
display(mejores_procedimientos_torax)

ANIO_EGRESO  COD_HOSPITAL procedimiento  conteo  ranking  \
175914         2021        112103         39.61     626        1   
175915         2021        112103         39.64     584        1   
175922         2021        112103         42.23     391        1   
175926         2021        112103         37.22     280        1   
175927         2021        112103         89.46     278        1   
...             ...           ...           ...     ...      ...   
34798          2019        112103         34.89       2        1   
34794          2019        112103         33.49       2        1   
34786          2019        112103         07.81       2        1   
34897          2019        112103         38.46       1        1   
34895          2019        112103         38.42       1        1   

        diferencia_proximo_hospital  total_procedimientos  
175914                           68                  3184  
175915                          246                  2020  
175922                          342                   607  
175926                           22                   801  
175927                           82                  1355  
...                             ...                   ...  
34798                             1                     6  
34794                             1                    10  
34786                             1                    10  
34897                             0                     3  
34895                             0                     5  

[265 rows x 7 columns]

Los resultados muestran que el tórax en el año 2021 fue altamente relevante en los procedimientos
39.61, 39.64, 42.23, entre otros. Además, se puede observar la diferencia que tiene el Tórax
con los siguientes establecimientos de salud y el total de procedimientos realizados en el país.

In [11]:
cie_9 = pd.read_excel("../data/external/CIE-9.xlsx", dtype={"Código": str})
esquema_hospitales = pd.read_excel(
    "../data/external/Esquema_Registro-2023.xls", sheet_name=4, header=7
).dropna(how="all", axis=1)

In [12]:
ranking_con_nombre_procedimientos = pd.merge(
    conteo_y_ranking_procedimientos, cie_9, how="left", left_on="procedimiento", right_on="Código"
)
ranking_con_nombre_procedimientos_y_establecimientos = ranking_con_nombre_procedimientos.merge(
    esquema_hospitales, how="left", left_on="COD_HOSPITAL", right_on="Código nuevo Establecimiento"
)

In [13]:
display(ranking_con_nombre_procedimientos_y_establecimientos)

ANIO_EGRESO  COD_HOSPITAL procedimiento  conteo  ranking  \
0              2021        113180         99.99       4        1   
1              2021        108100         99.99       2        2   
2              2021        103100         99.99       1        3   
3              2021        105100         99.99       1        4   
4              2021        110100         99.99       1        5   
...             ...           ...           ...     ...      ...   
214751         2019        105101         00.01      65        1   
214752         2019        106100         00.01      13        2   
214753         2019        113100         00.01       1        3   
214754         2019        120101         00.01       1        4   
214755         2019        114105             0       2        1   

        diferencia_proximo_hospital  total_procedimientos Código  \
0                                 2                    18  99.99   
1                                 1                    18  99.99   
2                                 0                    18  99.99   
3                                 0                    18  99.99   
4                                 0                    18  99.99   
...                             ...                   ...    ...   
214751                           52                    80  00.01   
214752                           12                    80  00.01   
214753                            0                    80  00.01   
214754                           -1                    80  00.01   
214755                         <NA>                     2    NaN   

                                         Descripción  \
0                      PROCEDIMIENTO MISCELANEO.OTRO   
1                      PROCEDIMIENTO MISCELANEO.OTRO   
2                      PROCEDIMIENTO MISCELANEO.OTRO   
3                      PROCEDIMIENTO MISCELANEO.OTRO   
4                      PROCEDIMIENTO MISCELANEO.OTRO   
...                                              ...   
214751  ULTRASONIDO TERAPEUTICO.VASO CABEZA Y CUELLO   
214752  ULTRASONIDO TERAPEUTICO.VASO CABEZA Y CUELLO   
214753  ULTRASONIDO TERAPEUTICO.VASO CABEZA Y CUELLO   
214754  ULTRASONIDO TERAPEUTICO.VASO CABEZA Y CUELLO   
214755                                           NaN   

                                                Categoría  ...  \
0                  99 OTROS PROCEDIMIENTOS NO QUIRURGICOS  ...   
1                  99 OTROS PROCEDIMIENTOS NO QUIRURGICOS  ...   
2                  99 OTROS PROCEDIMIENTOS NO QUIRURGICOS  ...   
3                  99 OTROS PROCEDIMIENTOS NO QUIRURGICOS  ...   
4                  99 OTROS PROCEDIMIENTOS NO QUIRURGICOS  ...   
...                                                   ...  ...   
214751  00 PROCEDIMIENTOS E INTERVENCIONES, NO CLASIFI...  ...   
214752  00 PROCEDIMIENTOS E INTERVENCIONES, NO CLASIFI...  ...   
214753  00 PROCEDIMIENTOS E INTERVENCIONES, NO CLASIFI...  ...   
214754  00 PROCEDIMIENTOS E INTERVENCIONES, NO CLASIFI...  ...   
214755                                                NaN  ...   

                                                 Capítulo  \
0       16. (87 99) PROCEDIMIENTOS DIAGNOSTICOS Y TERA...   
1       16. (87 99) PROCEDIMIENTOS DIAGNOSTICOS Y TERA...   
2       16. (87 99) PROCEDIMIENTOS DIAGNOSTICOS Y TERA...   
3       16. (87 99) PROCEDIMIENTOS DIAGNOSTICOS Y TERA...   
4       16. (87 99) PROCEDIMIENTOS DIAGNOSTICOS Y TERA...   
...                                                   ...   
214751  00. (00 00) PROCEDIMIENTOS E INTERVENCIONES NO...   
214752  00. (00 00) PROCEDIMIENTOS E INTERVENCIONES NO...   
214753  00. (00 00) PROCEDIMIENTOS E INTERVENCIONES NO...   
214754  00. (00 00) PROCEDIMIENTOS E INTERVENCIONES NO...   
214755                                                NaN   

       Código nuevo Establecimiento  \
0                            113180   
1                            108100   
2                            103100   
3                  

Ahora se puede observar el nombre del procedimiento analizado y el nombre del hospital visto.

In [17]:
ranking_con_nombre_procedimientos_y_establecimientos.to_csv(
    "../data/interim/procedimientos_nacionales_ranking.csv",
    sep=";",
    encoding="latin-1",
    index=False,
)